File: langs/php/tests/ValidatorTest.php
<?php

declare(strict_types=1);

namespace Dotlyte\Tests;

use Dotlyte\SchemaRule;
use Dotlyte\SchemaViolation;
use Dotlyte\ValidationException;
use Dotlyte\Validator;
use PHPUnit\Framework\TestCase;

final class ValidatorTest extends TestCase
{
    public function testValidData(): void
    {
        $schema = [
            'port' => new SchemaRule(type: 'integer', required: true),
            'host' => new SchemaRule(type: 'string', required: true),
        ];
        $data = ['port' => 8080, 'host' => 'localhost'];
        $violations = Validator::validate($data, $schema);
        $this->assertCount(0, $violations);
    }

    public function testMissingRequired(): void
    {
        $schema = [
            'port' => new SchemaRule(type: 'integer', required: true),
        ];
        $data = [];
        $violations = Validator::validate($data, $schema);
        $this->assertCount(1, $violations);
        $this->assertSame('port', $violations[0]->key);
        $this->assertStringContainsString('required', strtolower($violations[0]->message));
    }

    public function testTypeMismatch(): void
    {
        $schema = [
            'port' => new SchemaRule(type: 'integer'),
        ];
        $data = ['port' => 'not_a_number'];
        $violations = Validator::validate($data, $schema);
        $this->assertCount(1, $violations);
    }

    public function testEnumValidation(): void
    {
        $schema = [
            'env' => new SchemaRule(type: 'string', enumValues: ['dev', 'staging', 'prod']),
        ];
        $data = ['env' => 'dev'];
        $this->assertCount(0, Validator::validate($data, $schema));

        $data = ['env' => 'invalid'];
        $this->assertCount(1, Validator::validate($data, $schema));
    }

    public function testMinMax(): void
    {
        $schema = [
            'port' => new SchemaRule(type: 'integer', min: 1, max: 65535),
        ];
        $data = ['port' => 0];
        $this->assertCount(1, Validator::validate($data, $schema));

        $data = ['port' => 8080];
        $this->assertCount(0, Validator::validate($data, $schema));

        $data = ['port' => 99999];
        $this->assertCount(1, Validator::validate($data, $schema));
    }

    public function testPattern(): void
    {
        $schema = [
            'name' => new SchemaRule(type: 'string', pattern: '/^[a-z]+$/'),
        ];
        $data = ['name' => 'hello'];
        $this->assertCount(0, Validator::validate($data, $schema));

        $data = ['name' => 'Hello123'];
        $this->assertCount(1, Validator::validate($data, $schema));
    }

    public function testFormatEmail(): void
    {
        $schema = [
            'email' => new SchemaRule(type: 'string', format: 'email'),
        ];
        $data = ['email' => 'test@example.com'];
        $this->assertCount(0, Validator::validate($data, $schema));

        $data = ['email' => 'not-an-email'];
        $this->assertCount(1, Validator::validate($data, $schema));
    }

    public function testFormatUuid(): void
    {
        $schema = [
            'id' => new SchemaRule(type: 'string', format: 'uuid'),
        ];
        $data = ['id' => '550e8400-e29b-41d4-a716-446655440000'];
        $this->assertCount(0, Validator::validate($data, $schema));

        $data = ['id' => 'not-a-uuid'];
        $this->assertCount(1, Validator::validate($data, $schema));
    }

    public function testApplyDefaults(): void
    {
        $schema = [
            'port' => new SchemaRule(type: 'integer', defaultValue: 3000),
            'host' => new SchemaRule(type: 'string', defaultValue: 'localhost'),
        ];
        $data = ['port' => 8080];
        Validator::applyDefaults($data, $schema);
        $this->assertSame(8080, $data['port']);
        $this->assertSame('localhost', $data['host']);
    }

    public function testSensitiveKeys(): void
    {
        $schema = [
            'password' => new SchemaRule(type: 'string', sensitive: true),
            'host' => new SchemaRule(type: 'string'),
            'api_key' => new SchemaRule(type: 'string', sensitive: true),
        ];
        $sensitive = Validator::sensitiveKeys($schema);
        $this->assertContains('password', $sensitive);
        $this->assertContains('api_key', $sensitive);
        $this->assertNotContains('host', $sensitive);
    }

    public function testAssertValidThrows(): void
    {
        $schema = [
            'port' => new SchemaRule(type: 'integer', required: true),
        ];
        $this->expectException(ValidationException::class);
        Validator::assertValid([], $schema);
    }

    public function testStrictMode(): void
    {
        $schema = [
            'port' => new SchemaRule(type: 'integer'),
        ];
        $data = ['port' => 8080, 'extra' => 'not-in-schema'];

        // Non-strict: no error for extra keys
        $this->assertCount(0, Validator::validate($data, $schema));

        // Strict: error for extra keys
        $this->assertCount(1, Validator::validate($data, $schema, strict: true));
    }

    public function testNestedKeyValidation(): void
    {
        $schema = [
            'database.host' => new SchemaRule(type: 'string', required: true),
            'database.port' => new SchemaRule(type: 'integer', required: true),
        ];
        $data = ['database' => ['host' => 'localhost', 'port' => 5432]];
        $this->assertCount(0, Validator::validate($data, $schema));
    }
}

File: langs/java/src/test/java/dev/dotlyte/ValidatorTest.java
package dev.dotlyte;

import org.junit.jupiter.api.Test;

import java.util.*;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Tests for v2 schema validation.
 */
class ValidatorTest {

    @Test
    void requiredKeyMissing() {
        Map<String, Object> data = new LinkedHashMap<>();
        Map<String, SchemaRule> schema = new LinkedHashMap<>();
        schema.put("db.host", new SchemaRule().required(true));

        List<SchemaViolation> violations = Validator.validate(data, schema, false);
        assertEquals(1, violations.size());
        assertEquals("required", violations.get(0).getRule());
    }

    @Test
    void typeCheckString() {
        Map<String, Object> data = new LinkedHashMap<>();
        data.put("port", 8080);
        Map<String, SchemaRule> schema = new LinkedHashMap<>();
        schema.put("port", new SchemaRule().type("string"));

        List<SchemaViolation> violations = Validator.validate(data, schema, false);
        assertEquals(1, violations.size());
        assertEquals("type", violations.get(0).getRule());
    }

    @Test
    void typeCheckNumber() {
        Map<String, Object> data = new LinkedHashMap<>();
        data.put("port", 8080);
        Map<String, SchemaRule> schema = new LinkedHashMap<>();
        schema.put("port", new SchemaRule().type("number"));

        List<SchemaViolation> violations = Validator.validate(data, schema, false);
        assertTrue(violations.isEmpty());
    }

    @Test
    void formatEmail() {
        Map<String, Object> data = new LinkedHashMap<>();
        data.put("email", "not-an-email");
        Map<String, SchemaRule> schema = new LinkedHashMap<>();
        schema.put("email", new SchemaRule().format("email"));

        List<SchemaViolation> violations = Validator.validate(data, schema, false);
        assertEquals(1, violations.size());
        assertEquals("format", violations.get(0).getRule());
    }

    @Test
    void formatEmailValid() {
        Map<String, Object> data = new LinkedHashMap<>();
        data.put("email", "user@example.com");
        Map<String, SchemaRule> schema = new LinkedHashMap<>();
        schema.put("email", new SchemaRule().format("email"));

        assertTrue(Validator.validate(data, schema, false).isEmpty());
    }

    @Test
    void enumCheck() {
        Map<String, Object> data = new LinkedHashMap<>();
        data.put("env", "staging");
        Map<String, SchemaRule> schema = new LinkedHashMap<>();
        schema.put("env", new SchemaRule().enumValues(Arrays.asList("dev", "prod")));

        List<SchemaViolation> violations = Validator.validate(data, schema, false);
        assertEquals(1, violations.size());
        assertEquals("enum", violations.get(0).getRule());
    }

    @Test
    void minMaxValidation() {
        Map<String, Object> data = new LinkedHashMap<>();
        data.put("port", 100);
        Map<String, SchemaRule> schema = new LinkedHashMap<>();
        schema.put("port", new SchemaRule().min(1024.0).max(65535.0));

        List<SchemaViolation> violations = Validator.validate(data, schema, false);
        assertEquals(1, violations.size());
        assertEquals("min", violations.get(0).getRule());
    }

    @Test
    void strictModeRejectsUnknown() {
        Map<String, Object> data = new LinkedHashMap<>();
        data.put("port", 8080);
        data.put("extra", "value");
        Map<String, SchemaRule> schema = new LinkedHashMap<>();
        schema.put("port", new SchemaRule().type("number"));

        List<SchemaViolation> violations = Validator.validate(data, schema, true);
        assertEquals(1, violations.size());
        assertEquals("strict", violations.get(0).getRule());
    }

    @Test
    void applyDefaults() {
        Map<String, Object> data = new LinkedHashMap<>();
        Map<String, SchemaRule> schema = new LinkedHashMap<>();
        schema.put("port", new SchemaRule().defaultValue(3000));

        Validator.applyDefaults(data, schema);
        assertEquals(3000, data.get("port"));
    }

    @Test
    void assertValidThrows() {
        Map<String, Object> data = new LinkedHashMap<>();
        Map<String, SchemaRule> schema = new LinkedHashMap<>();
        schema.put("required_key", new SchemaRule().required(true));

        assertThrows(DotlyteException.class, () ->
                Validator.assertValid(data, schema, false));
    }

    @Test
    void sensitiveKeysExtraction() {
        Map<String, SchemaRule> schema = new LinkedHashMap<>();
        schema.put("db.password", new SchemaRule().sensitive(true));
        schema.put("db.host", new SchemaRule());

        List<String> sensitive = Validator.getSensitiveKeys(schema);
        assertEquals(1, sensitive.size());
        assertEquals("db.password", sensitive.get(0));
    }

    @Test
    void patternValidation() {
        Map<String, Object> data = new LinkedHashMap<>();
        data.put("code", "ABC");
        Map<String, SchemaRule> schema = new LinkedHashMap<>();
        schema.put("code", new SchemaRule().pattern("^[0-9]+$"));

        List<SchemaViolation> violations = Validator.validate(data, schema, false);
        assertEquals(1, violations.size());
        assertEquals("pattern", violations.get(0).getRule());
    }
}
